cudamapper/src/main.cu

```diff
@@ -105,148 +118,171 @@ int main(int argc, char* argv[])
+
+        std::int32_t target_start_index = 0;
+        // If all_to_all mode, then we can optimzie by starting the target sequences from the same index as
+        // query because all indices before the current query index are guaranteed to have been processed in
+        // a2a mapping.
+        if (all_to_all)
         {
-            overlaps_to_write.pop_back();
+            target_start_index = query_start_index;
         }
-        overlaps_writer_mtx.unlock();
-    };
 
-    // Start async thread for writing out PAF
-    auto overlaps_writer_func = [&overlaps_to_write, &overlaps_writer_mtx]() {
-        while (true)
+        for (; target_start_index < targets; target_start_index += target_index_size)
         {
-            bool done = false;
-            overlaps_writer_mtx.lock();
-            if (!overlaps_to_write.empty())
-            {
-                CGA_NVTX_RANGE(profile, "overlaps_writer");
-                std::vector<claragenomics::cudamapper::Overlap>& overlaps = overlaps_to_write.front();
-                // An empty overlap vector indicates end of processing.
-                if (overlaps.size() > 0)
-                {
-                    claragenomics::cudamapper::Overlapper::print_paf(overlaps);
-                    overlaps_to_write.pop_front();
-                    overlaps_to_write.shrink_to_fit();
-                }
-                else
-                {
-                    done = true;
-                }
-            }
-            overlaps_writer_mtx.unlock();
-            if (done)
+            std::int32_t target_end_index = std::min(target_start_index + target_index_size,
+                                                     static_cast<size_t>(targets));
+            q.target_ranges.push_back(std::make_pair(target_start_index, target_end_index));
+        }
+
+        query_target_ranges.push_back(q);
+    }
+
+    // This is a per-device cache, if it has the index it will return it, if not it will generate it, store and return it.
+    std::vector<std::map<std::pair<uint64_t, uint64_t>, std::shared_ptr<claragenomics::cudamapper::Index>>> index_cache(num_devices);
+
+    auto get_index = [&index_cache, &max_cache_size](claragenomics::io::FastaParser& parser,
+                                                     const claragenomics::cudamapper::read_id_t query_start_index,
+                                                     const claragenomics::cudamapper::read_id_t query_end_index,
+                                                     const std::uint64_t k,
+                                                     const std::uint64_t w,
+                                                     int device_id) {
+        std::pair<uint64_t, uint64_t> key;
+        key.first  = query_start_index;
+        key.second = query_end_index;
+
+        std::shared_ptr<claragenomics::cudamapper::Index> index;
+
+        if (index_cache[device_id].count(key))
+        {
+            index = index_cache[device_id][key];
```

Comment:
Very good point. Caching doesn't even make sense int he query-target situation. I have added a check so that entries are only added to the cache in the case of all_to_all.

Reply:
Why does it not make sense? Let's say query is split into chunks/indices q1, q2, q3 and target in t1, t2, t3, t4, t5. Without chunking t1, t2, t3, t4, t5 would be generated three times (once for each query chunk)

Reply:
You are right, the caching strategy we are using here however wouldn't make sense, - in a "standard" query vs target matching we should only ever cache the target reads and there should be no eviction from the cache at all (since all query indices are used only once and all target indices are used the same number of times).